Harmonize exit codes of the imaging scripts

cmadjar opened the pull request:
In this PR, a NeuroDB/ExitCodes.pm class was created. It lists all the exit codes currently in use in the insertion scripts and harmonize them.

The exit codes are organized per script, allowing 10 possible exit codes per script (note: not all of the possible exit codes are used by each script, giving some room to add some later on if needed). 

For each script, exit codes are organized based on their use (validation failures, database related failures, file related failures, script execution failures, study related failures, input error checking and setting failures).

From now on, when exiting a script, it should be calling a variable that is initialized in the ExitCodes.pm library. For example, upon successful execution of the script, should exit the script as follows:
`exit $NeuroDB::ExitCodes::SUCCESS;`

**Note for the release**: most of the error codes will be different now (e.g in DTIPrep, when the XML protocol could not be read we used to exit with error code 33, but now it's $NeuroDB::ExitCodes::UNREADABLE_DTIPREP_PROTOCOL = 60). Since the ServerProcessesManager displays these error codes, it might be worth mentioning (as a caveat for existing project) that their meaning has changed now. This should affect only projects with using the MRI upload auto launch.

Redmine ticket associated to this PR: https://redmine.cbrain.mcgill.ca/issues/9966

Below is a list of the possible exit codes organized per script:

1. Common exit codes to most insertion scripts (exit codes from 0 to 19, 0 = exit script with success status)
2. Exit codes from batch_uploads_imageuploader (exit codes from 20 to 39)
3. Exit codes from batch_uploads_tarchive (no exit codes available yet, exit codes will be from 40 to 59)
4. Exit codes from dicom-archive/dicomTar.pl (exit codes from 60 to 79)
5. Exit codes from dicom-archive/updateMRI_upload (exit codes from 80 to 99)
6. Exit codes from DTIPrep/DTIPrep_pipeline.pl (exit codes from 100 to 119)
7. Exit codes from DTIPrep/DTIPrepRegister.pl (exit codes from 120 to 139)
8. Exit codes from uploadNeuroDB/imaging_upload_file.pl (exit codes from 140 to 159)
9. Exit codes from uploadNeuroDB/NeuroDB/ImagingUpload.pm (exit codes from 160 to 179)
10. Exit codes from uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm (exit codes  from 180 to 199)
11. Exit codes from uploadNeuroDB/minc_deletion.pl (exit codes from 200 to 219)
12. Exit codes from uploadNeuroDB/minc_insertion.pl (exit codes from 220 to 239)
13. Exit codes from uploadNeuroDB/register_processed_data.pl (exit codes from
240 to 259)
14. Exit codes from uploadNeuroDB/tarchiveLoader (exit codes from 260 to 279)

cmadjar commented:
@nicolasbrossard I asked for your review since it was your suggestion to create this class last week. Thanks!!

MounaSafiHarab commented:
@cmadjar 
holy cecile! that is super fast!!!!

nicolasbrossard commented:
Beast.....

cmadjar commented:
@nicolasbrossard wait a bit before testing the PR. The beast realized early this morning that she did not test the scripts yet. I'm going to do that this morning and I will let you know once it is ready for testing. Thanks!!

cmadjar commented:
@nicolasbrossard Ready for testing!!

Note, I tested the execution of the following scripts and they all work:
- dicomTar.pl (which calls updateMRI_Upload.pl)
- tarchiveLoader (which calls tarchive_validation.pl, minc_insertion.pl, mass_nii.pl, mass_jiv.pl, mass_pic.pl and MRIProcessingUtility.pm)
- batch_uploads_imaginguploader (which calls imaging_upload_file.pl and ImagingUpload.pm, on top of the scripts already mentioned above)
- DTIPrep_pipeline.pl (which calls DTIPrepRegister.pl and register_processed_data.pl)

That should cover everything.

cmadjar commented:
@nicolasbrossard Thank you for the suggestion! I added the 'print STDERR' when exiting due to an error. I hope I did not miss any. :S

nicolasbrossard commented on uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm:
```diff
@@ -172,7 +174,7 @@ sub extract_tarchive {
         print $message;
```
Missing STDERR. There is also another one just like this on line 74.

cmadjar commented:
@nicolasbrossard Thanks! I corrected the two that I missed. Additionally, I realized the exit codes were used by writeErrorLog() so I replaced the hardcoded ones by the one set in ExitCodes.pm

MounaSafiHarab commented:
@cmadjar @nicolasbrossard 

one thing that can benefit from this change in the front-end is the server_process_manager module which currently displays the exit code itself, we can add a column to make it display the meaning of that exit code (or simply replace the existing exit code column with the exit meaning column); much more meaningful to the user...

cmadjar commented:
@MounaSafiHarab @nicolasbrossard is the imaging pipeline using the server_processes table? Mine is completely empty... I also tried to grep for it in the LORIS-MRI code and could not find anything. I'm confused :S

MounaSafiHarab commented:
> @MounaSafiHarab @nicolasbrossard is the imaging pipeline using the server_processes table? Mine is completely empty... I also tried to grep for it in the LORIS-MRI code and could not find anything. I'm confused :S

No, only if you auto-launch Imaging Uploader would this table be populated. It is not on the MRI side, but rather on the LORIS side. My earlier comment was just an improvement that we can do in Loris to make use of/build up on your exit code changes

cmadjar commented:
Aaaah. That would be a really cool improvement indeed!! Thanks for the clarification!

nicolasbrossard commented:
@cmadjar @MounaSafiHarab Agreed. I opened a Redmine ticket for that so we don't forget:

https://redmine.cbrain.mcgill.ca/issues/13941

nicolasbrossard commented:
I approved the changes. If I am not mistaken, some (all?) error codes will be different now  (e.g in DTIPrep, when the XML protocol could not be read we used to exit with error code 33, but now it's `$NeuroDB::ExitCodes::UNREADABLE_DTIPREP_PROTOCOL` = 60). Since the ServerProcessesManager displays these error codes, it might be worth mentioning in the Release Notes that their meaning has changed now. No biggie, as I think MAVAN was the only project using MRI upload auto launch...

cmadjar commented:
@nicolasbrossard You're right, I added the flag and copied your comment to add to the release notes. As you mention, this will probably affect only MAVAN which is not scanning anymore if I am not mistaken. Good timing ;)

MounaSafiHarab commented:
@cmadjar
can you please change the base branch to minor? just so that we make sure we do not merge accidentally into 19.0-dev

cmadjar commented:
> @cmadjar
can you please change the base branch to minor? just so that we make sure we do not merge accidentally into 19.0-dev

@MounaSafiHarab Just changed the base branch to minor. Thanks!!

MounaSafiHarab commented on uploadNeuroDB/NeuroDB/ExitCodes.pm:
```diff
@@ -0,0 +1,321 @@
+our $PHANTOM_ENTRY_FAILURE      = 11; # if the phantom entry in the text file is
+                                      # not 'N' nor 'Y'
+our $PNAME_FILENAME_MISMATCH    = 12; # if patient name and beginning of
+                                      # uploaded filename does not match
```
`do not match` instead of `does not match`

cmadjar replied:
oups... Fixed now.

MounaSafiHarab commented on uploadNeuroDB/register_processed_data.pl:
```diff
@@ -142,7 +151,7 @@
     if  (!$psc)     { 
         print LOG "\nERROR: No center found for this candidate \n\n"; 
-        exit 77; 
+        exit $NeuroDB::ExitCodes::GET_PSC_FROM_SOURCEFILEID_FAILURE;
```
why not use the already defined `$GET_PSC_FAILURE` defined as exit code 91 (from MRIProcessingUtility.pm)? it seems to be the same error to me; don't you think?

cmadjar replied:
Good point. I was actually on the fence about this one when I created it but could not decide ;). Will be using $GET_PSC_FAILURE

MounaSafiHarab commented on uploadNeuroDB/register_processed_data.pl:
```diff
@@ -171,8 +180,9 @@
-    exit 2;
+    print LOG "\nERROR: could not determine scannerID based on sourceFileID "
+              . "$sourceFileID.\n\n";
+    exit $NeuroDB::ExitCodes::GET_SCANNERID_FROM_SOURCEFILEID_FAILURE;
```
same comment here as above (why not use the already existing $GET_SCANNERID_FAILURE)?

cmadjar replied:
done

MounaSafiHarab commented on uploadNeuroDB/NeuroDB/ExitCodes.pm:
```diff
@@ -0,0 +1,321 @@
+
+# database related failure
+# called from minc_insertion.pl & register_processed_data.pl
+our $FILE_NOT_UNIQUE = 6; # if file to register is not unique & already
```
don't you think that leaving a range of 9 exits per category might be too tight?;
if we add 3 new ones here, what do we do next? 
maybe a minor thing, but thought of raising it anyway

cmadjar replied:
Good point! Will modify it to maybe 20 exit codes instead of 10 per category

MounaSafiHarab replied:
I would think 20 is good;
what does @nicolasbrossard think?
I know he had mentioned to me that we try to stay away from the 1000 range because MySQL uses those when it issues errors; but I think we are far from those...

nicolasbrossard replied:
Yeah, I think 20 is good. Let's just try to be reasonable regarding the number of error codes that we define for a given script: for example if a script parses an input file and we know that this operation can fail for multiple reasons, we might get away with defining a single error code $PARSING_FAILED instead of $PARSING_FAILED_INVALID SYNTAX, $PARSING_FAILED_DUPLICATE_IDS, etc...

MounaSafiHarab commented:
still going through the scripts one at a time;
but a few comments here:

1) the installer script has exit codes as well. Clearly we can not rely on your new class because the user did not install Loris-MRI yet ;) but the exit codes there are numbered similar to some of our exit codes (in the common category of 1-10), could that be confusing?

2) none of the scripts in the tools/ directory have been changed; is this intentional? I did not expect you to change anything, but if we are to do this later, should we create a redmine task about it? especially that some exit codes in those scripts are identical to other exit codes with different meaning now... but one can argue these scripts are only to be run at release time fr past releases... so either way is fine for me

3) the scripts in MNI/ and inside uploadNeuroDB/bin/  as well as: undoUpload, database_files_update, and CleanupTarchive; are we leaving those out?

MounaSafiHarab commented on uploadNeuroDB/tarchiveLoader:
```diff
@@ -183,10 +189,9 @@ unless ($tarchive =~ m/$tarchivePath/i) {
-    exit 4;
+    print STDERR "\nERROR: Could not find archive $tarchive.\n"
+                 . "Please, make sure the path to the archive is correct.\n\n";
+    exit $NeuroDB::ExitCodes::ARG_FILE_DOES_NOT_EXIST;
```
@cmadjar 
you forgot ONE exit code on success to change it :) (yes only ONE)
please scroll to line 251 (249 in original script before your changes)... I can not comment there so I am putting it here instead

cmadjar replied:
Good catch!!

MounaSafiHarab commented on DTIPrep/DTIPrepRegister.pl:
```diff
@@ -181,7 +191,7 @@
 if (!$registeredXMLprotocolID) {
     print LOG "\nERROR: no XML protocol file was registered in the database\n";
-    exit 0;
+    exit $NeuroDB::ExitCodes::XML_PROTOCOL_INSERT_FAILURE;
```
mmm, just curious, isn't exit 0 reserved for Success???

cmadjar replied:
Indeed, which is why it was a mistake in the past to exit 0 when the file registration failed. I created the script and did not know the meaning of the exit code at the time, so I would not really rely on what I wrote at that time ;).

MounaSafiHarab commented on DTIPrep/DTIPrep_pipeline.pl:
```diff
@@ -58,26 +59,31 @@
 if (!$DTIPrepVersion) {
-    print "$Usage\n\t ERROR: Pipeline version could not been determined via the path to DTIPrep binary. You need to specify which version of DTIPrep you will be using with -version option.\n\n";  
-    exit 33;
+    print STDERR "$Usage\n\t ERROR: Pipeline version could not been determined "
```
grammar error "could not been determined" :)

cmadjar replied:
oups...

MounaSafiHarab commented on docs/scripts_md/ExitCodes.md:
```diff
@@ -0,0 +1,74 @@
+
+The exit codes are organized per script, allowing 10 exit codes per script
+(note, not all of the possible exit codes are used by each script, giving
+some room to add some later on if needed). For each scripts, exit codes are
```
script not scripts

cmadjar replied:
done

MounaSafiHarab commented on docs/scripts_md/ExitCodes.md:
```diff
@@ -0,0 +1,74 @@
+
+This class lists all the exit codes used by the imaging insertion scripts.
+
+The exit codes are organized per script, allowing 10 exit codes per script
```
you cna say they are organized per script, together with a section that represents common failures to most scripts.

cmadjar replied:
👍

MounaSafiHarab commented on DTIPrep/DTIPrepRegister.pl:
```diff
@@ -206,7 +216,7 @@
 if (!$registeredXMLReportFile) {
     print LOG "\nERROR: no XML report file was registered in the database\n";
-    exit 0;
+    exit $NeuroDB::ExitCodes::XML_QCREPORT_INSERT_FAILURE;
```
also, why not use the code `$FILE_REGISTRATION_FAILURE`?

cmadjar replied:
Good point! Will be changing those.

MounaSafiHarab commented on uploadNeuroDB/mass_jiv.pl:
```diff
@@ -30,17 +32,21 @@
+if ( !@Settings::db ) {
+    print STDERR "\n\tERROR: You don't have a \@db setting in the file "
+                 . "$ENV{LORIS_CONFIG}/.loris_mri/$profile \n\n";
+    exit $NeuroDB::ExitCodes::DB_SETTINGS_FAILURE;}
```
just out of curiosity, what if the file `prod` does not exist?
we only check if the argument is given and then if the database section is in the prod file, but never see if the prod file is there; as is done here on line 42 above (`if (-f ...)')

cmadjar replied:
yeah. That is bad... I see it is like that in most scripts. :S Should I change it in this PR or create a redmine ticket to change this in another PR? What do you think is the best path? I could do either :).

MounaSafiHarab replied:
no it could be done in a separate one; otherwise this PR would be fixing everything wrong in the MRI scripts and you won't finish anytime soon:) just wanted to raise this comment as I saw it when reviewing the changes...

cmadjar commented:
> 1. get_dicom_info.pl: on line 132, there is this exit($status); shouldn't this be fixed? (check the TwoArgInfoOption() in that script)

Good catch!! Will do the change shortly. Have a meeting to go to...

> 2. the installer script has exit codes as well. Clearly we can not rely on your new class because the user did not install Loris-MRI yet ;) but the exit codes there are numbered similar to some of our exit codes (in the common category of 1-10), could that be confusing?

I could at least harmonize the exit code number in the imaging install script. It does not hurt to do so. 

> 3. none of the scripts in the tools/ directory have been changed; is this intentional? I did not expect you to change anything, but if we are to do this later, should we create a redmine task about it? especially that some exit codes in those scripts are identical to other exit codes with different meaning now... but one can argue these scripts are only to be run at release time fr past releases... so either way is fine for me

So far I only touched the main scripts for this PR. Maybe we could do the tools script in another one? 

> 4. the scripts in MNI/ and inside uploadNeuroDB/bin/ as well as: undoUpload, database_files_update, and CleanupTarchive; and others that I don't think I ever heard of!, are we leaving those out?

For the scripts in MNI/, I don't think we (as in LORIS) are the ones that wrote them (I think they got imported somehow from the minc tools? Therefore, I did not feel confident modifying those libraries.

For the uploadNeuroDB/bin, I missed them... I'll make the change for those

MounaSafiHarab commented:
> For the uploadNeuroDB/bin, I missed them... I'll make the change for those

I think this and the answer to point 2) above can be done in another PR. I personally do not mind this being done in separate steps, especially as the main scripts have been changed here.
(and I, too, do not really know why we have those duplicate MNI/ directory that are also in the MINC toolkit... I will ask this question in the next Imaging meeting).

MounaSafiHarab commented on uploadNeuroDB/tarchiveLoader:
```diff
@@ -248,7 +248,7 @@ if ($xlog) {
     if ($childPID == 0) { 
         exec("xterm -geometry 130x70 -e tail -f $logfile"); 
-        exit(0); 
+        exit $NeuroDB::ExitCodes::XLOG_FAILURE;
```
why is exit (0) now a failure?
I have to admit I am not sure I understand this `if` clause...

MounaSafiHarab commented on docs/scripts_md/ExitCodes.md:
```diff
@@ -21,12 +21,13 @@ insertion scripts
-file related failures, script execution failures, study related failures,
-input error checking and setting failures).
+The exit codes are organized per script, together with a section that
+represents common failures to most scripts. (note, not all of the possible
```
(very minor comment): Perhaps best to put the Note in the end as it distracts from the description.

cmadjar commented:
> I think this and the answer to point 2) above can be done in another PR. I personally do not mind this being done in separate steps, especially as the main scripts have been changed here.
(and I, too, do not really know why we have those duplicate MNI/ directory that are also in the MINC toolkit... I will ask this question in the next Imaging meeting).

I created two redmine tickets to do the scripts in tools and uploadNeuroDB/bin ([#14055](https://redmine.cbrain.mcgill.ca/issues/14055) and [14054)](https://redmine.cbrain.mcgill.ca/issues/14054)

cmadjar commented on uploadNeuroDB/tarchiveLoader:
```diff
@@ -243,7 +248,7 @@ if ($xlog) {
     if ($childPID == 0) { 
         exec("xterm -geometry 130x70 -e tail -f $logfile"); 
-        exit(0); 
+        exit $NeuroDB::ExitCodes::SUCCESS;
```
@nicolasbrossard Could you check that I understood correctly this code and that it should indeed return success here?

nicolasbrossard replied:
If `exec` runs the command successfully, it *never* returns. So whatever you put after `exec` will never be executed. You should remove the `exit....` line altogether. Also, you usually issue a call to `exec` like this:

`exec("my_command my_arg1 my_arg_2") or die "Command my_command my_arg1 my_arg2 failed: $!\n"`;

cmadjar commented:
@MounaSafiHarab @nicolasbrossard Ready for final review :)

MounaSafiHarab commented:
> @MounaSafiHarab @nicolasbrossard Ready for final review :)

I will 
thanks @cmadjar for all the changes
and @nicolasbrossard  for the clarification on that 'useless' log file exit code!

MounaSafiHarab commented on uploadNeuroDB/NeuroDB/ExitCodes.pm:
```diff
@@ -0,0 +1,314 @@
+
+#### --- FROM batch_uploads_tarchive
+
+#TODO: TO BE DONE LATER ON (ERROR CODES WOULD BE BETWEEN 20 & 29)
```
I think these numbers (20&29) are no longer valid...

MounaSafiHarab commented on uploadNeuroDB/NeuroDB/ExitCodes.pm:
```diff
@@ -0,0 +1,314 @@
+1. Common exit codes to most insertion scripts (exit codes from 0 to 19, 0 =
+exit script with success status)
+
+2. Exit codes from batch_uploads_imageuploader (exit codes from 20 to 49)
```
Is this intentional 20 to 49 or you meant 20-39?
I don;t care either way... but I have a feeling you intended 39 instead of 49.

MounaSafiHarab commented on uploadNeuroDB/NeuroDB/ExitCodes.pm:
```diff
@@ -0,0 +1,314 @@
+3. Exit codes from batch_uploads_tarchive (no exit codes available yet, exit
+codes will be from 60 to 69)
+
+4. Exit codes from dicom-archive/dicomTar.pl (exit codes from 80 to 89)
```
on line 144 in this file, this starts at exit codes 40 not 80;

MounaSafiHarab commented on uploadNeuroDB/NeuroDB/ExitCodes.pm:
```diff
@@ -0,0 +1,314 @@
+
+4. Exit codes from dicom-archive/dicomTar.pl (exit codes from 80 to 89)
+
+5. Exit codes from dicom-archive/updateMRI_upload (exit codes from 100 to 119)
```
same comment here (on line 161 this is code 60 as a starting point not 100)

MounaSafiHarab commented:
@cmadjar 

1) I added a few more comments (mainly that the numbers inside the ExitCode.pm do not match in the description and the variable declaration; same thing applies to the main description of this pull request).

And

2) sorry now after one PR was merged, you need to rebase :( there is conflict

cmadjar commented:
@MounaSafiHarab Good catches! The numbering should be resolved now. If not, I really need to learn how to count again ;). Resolved the conflict and updated the description as well. Ready for another round of review!

MounaSafiHarab commented on uploadNeuroDB/minc_insertion.pl:
```diff
@@ -461,7 +462,7 @@
                    'minc_insertion.pl', $upload_id, 'Y', 
                    $notify_notsummary);
-   exit 9;
+   exit $NeuroDB::ExitCodes::UNKNOW_PROTOCOL;
```
being picky, but this should be UNKNOWN (missing N in the end)

MounaSafiHarab commented:
@cmadjar 

I did another test today where I commented out the `db` section from my prod file;
on the terminal, I get this error `ERROR: You don't have a @db setting in the file /data/loris-MRI/bin/mri/dicom-archive/.loris_mri/prod` as expected. I expected an exit code that corresponds to `DB_SETTINGS_FAILURE`; i.e. 4, but the exit code it shows in the server_process_manager is 2 `PROFILE_FAILURE` which I think is wrong... 

@nicolasbrossard 
can you please test? to see if you can reproduce my error. I uploaded a scan, put auto launch on, and commented out the `@db` section in the prod file. I get exit error code 2. If I launch this at the command line, I get the error message that would have to correspond to exit code 4 not 2...

MounaSafiHarab commented:
> @cmadjar
> 
> I did another test today where I commented out the db section from my prod file;
> on the terminal, I get this error ERROR: You don't have a @db setting in the file /data/loris-MRI/bin/mri/dicom-archive/.loris_mri/prod as expected. I expected an exit code that corresponds to DB_SETTINGS_FAILURE; i.e. 4, but the exit code it shows in the server_process_manager is 2 PROFILE_FAILURE which I think is wrong...
> 
> @nicolasbrossard
> can you please test? to see if you can reproduce my error. I uploaded a scan, put auto launch on, and commented out the @db section in the prod file. I get exit error code 2. If I launch this at the command line, I get the error message that would have to correspond to exit code 4 not 2...

sat with @nicolasbrossard re. this issue, it turns out something unrelated in the imaging uploader when auto launch is on. so a redmine ticket is created about it here: RM#14093
